pkg/crawler/crawler.go

```diff
@@ -82,334 +128,474 @@ func NewCrawler(opt Option) (Crawler, error) {
+	if groupID == "" || artifactID == "" {
+		return nil
 	}
 
-	if len(foundVersions) == 0 {
+	// Skip if already processed
+	// NOTE: We only need to check if this GAV has been processed before,
+	// no need to store anything as the map is pre-populated
+	gavHash := hash.GAV(groupID, artifactID, versionDir)
+	if _, exists := f.storedGAVs[gavHash]; exists {
 		return nil
 	}
 
-	slices.SortFunc(foundVersions, func(a, b types.Version) int {
-		return strings.Compare(a.Version, b.Version)
-	})
+	// Use the provided Maven client
+	sha1, err := f.client.FetchSHA1(ctx, item)
+	if sha1 == "N/A" {
+		f.errCount.Add(1)
+	} else if err != nil {
+		f.logger.Warn("Failed to fetch SHA1", slog.String("item", item), slog.Any("error", err))
+		return nil
+	}
 
-	index := &Index{
-		GroupID:     groupID,
-		ArtifactID:  artifactID,
-		Versions:    foundVersions,
-		ArchiveType: types.JarType,
+	// Send record to next stage
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	case recordCh <- Record{ // Record sent to channel
+		GroupID:    groupID,
+		ArtifactID: artifactID,
+		VersionDir: versionDir,
+		Version:    version,
+		SHA1:       sha1,
+	}:
 	}
-	fileName := fmt.Sprintf("%s.json", index.ArtifactID)
-	filePath := filepath.Join(c.dir, index.GroupID, fileName)
-	if err = fileutil.WriteJSON(filePath, index); err != nil {
-		return xerrors.Errorf("json write error: %w", err)
+	return nil
+}
+
+// Run processes records and writes them to appropriate shard files
+func (a *Aggregator) Run(recordsCh <-chan Record) error {
+	a.logger.Info("Starting record aggregator")
+
+	// Close all writers and files when done
+	defer a.closeWriters()
+
+	for rec := range recordsCh {
+		// Calculate shard index based on GroupID+ArtifactID
+		shardIdx := int(hash.GA(rec.GroupID, rec.ArtifactID) % uint64(a.shardCount))
+
+		// Get or create writer for this shard
+		writer, err := a.newWriter(shardIdx)
+		if err != nil {
+			a.logger.Error("Failed to get writer",
+				slog.Int("shard", shardIdx),
+				slog.Any("error", err))
+			return xerrors.Errorf("failed to get writer: %w", err)
+		}
+
+		// Since versionDir is the same as version in most cases, we can exclude it from the record for saving space.
+		if rec.VersionDir == rec.Version {
+			rec.Version = "-"
+		}
+
+		// Write TSV record: GroupID, ArtifactID, Version, SHA1
+		err = writer.Write([]string{
+			rec.GroupID,
+			rec.ArtifactID,
+			rec.VersionDir,
+			rec.Version,
+			rec.SHA1,
+		})
+		if err != nil {
+			a.logger.Error("Failed to write record",
+				slog.Int("shard", shardIdx),
+				slog.Any("error", err))
+			return xerrors.Errorf("failed to write record: %w", err)
+		}
+
+		// Periodically flush to disk
+		a.recordsProcessed++
+		if a.recordsProcessed%100000 == 0 {
+			// Flush all writers
+			a.flushWriters()
+			a.logger.Info(fmt.Sprintf("Processed %d records", a.recordsProcessed))
+		}
 	}
+
+	// Final flush of CSV writers and buffer writers
+	a.flushWriters()
```

Comment:
NIt:
You write final buffers twice:
- here 
- https://github.com/aquasecurity/trivy-java-db/blob/9fdf0b0c524d465b1a0be317babdadad2377d70f/pkg/crawler/crawler.go#

Reply:
The line number in the permalink seems to be missing.

Reply:
hm... sorry:
https://github.com/aquasecurity/trivy-java-db/blob/9fdf0b0c524d465b1a0be317babdadad2377d70f/pkg/crawler/crawler.go#L579

Reply:
Done
https://github.com/aquasecurity/trivy-java-db/pull/57/commits/50591d20ffd4f66e5fd443da852b39bd943083f5